### ISSUE TITLE
Does creating more tokens affects performance

### ISSUE THREAD

**Issue by mohansella** [cameron314/concurrentqueue#180]
Hi Cameron, like many users, i am having limited consumers (fixed threads with each one having a consumer token) and en-queue can be called from any thread. The readme section noted that calling with token will have better throughput. Since my application needs to be very high performant (in Millions/Sec), i am thinking about making tokens to all the possible producers which may count from 10 to 50 tokens from 10 threads. So, to the straight question, does creating more producer tokens will perform better than without tokens?

**Comment by cameron314** [comment]
It depends. The internal architecture of the queue uses independent sub-queues for each producer. Every producer token is registered as an individual sub-queue. This consumes memory (once a producer uses a block of memory, it is never let go until the producer is destroyed) and increases the number of queues that a consumer needs to check for items in the worst case (when the queue is empty).

If you do not use producer tokens, a sub-queue is created for each thread that is used to enqueue with. This is slightly slower, but not greatly so.

In your case, you seem to have significantly more producers than threads, so it may be better to use implicit producers (no producer tokens); this would take less memory. On the other hand, if you know that the queue is generally not going to be empty, it could be more efficient to use many producer tokens (and thus more sub-queues) since the consumer threads will have less contention pulling from individual sub-queues. This depends on the number of consumers as well.

Above all, the most important thing for performance is to use consumer tokens, which you are already doing.

So to sum up:
1) It's best to measure for your specific use case
2) Producer tokens may be better but bring a memory overhead if you have significantly more of them than threads
3) Implicit producers are slower than using a producer token, but not greatly so

**Comment by mohansella** [comment]
Thanks @cameron314 for the details. I will probably go with implicit producers. Anyway, I will update the findings once i complete my implementation (File Queue using this queue).
